contracts/MinimalBalance.sol

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.20;
+
+import {Currency, CurrencyLibrary} from "./types/Currency.sol";
+import {IMinimalBalance} from "./interfaces/IMinimalBalance.sol";
+
+/// An intentionally barebone balance mapping only supporting mint/burn/transfer
+contract MinimalBalance is IMinimalBalance {
+    using CurrencyLibrary for Currency;
+    // Mapping from Currency id to account balances
+
+    mapping(uint256 => mapping(address => uint256)) public balances;
+
+    /// @inheritdoc IMinimalBalance
+    function balanceOf(address account, Currency currency) public view returns (uint256) {
+        return balances[currency.toId()][account];
+    }
+
+    /**
+     * @notice Mint `amount` of currency `id` to `to`
+     *     @param to The address to mint to
+     *     @param id The id to mint
+     *     @param amount The amount to mint
+     */
+    function _mint(address to, uint256 id, uint256 amount) internal {
+        balances[id][to] += amount;
+        emit Mint(to, id, amount);
+    }
+
+    /**
+     * @notice Burn `amount` of currency `id` from `msg.sender`
+     *     @param id The id of the currency to burn
+     *     @param amount The amount to burn
+     *  @dev Will revert if the sender does not have enough balance
+     */
+    function _burn(uint256 id, uint256 amount) internal {
+        if (amount > balances[id][msg.sender]) revert InsufficientBalance();
```

Comment:
Am I missing something?
For this setup the user has to have transferred their tokens into the router, because the `msg.sender` is going to be the router. However theres no `transferFrom`... so the transfer of tokens would have to be in a different tx to the burn? And then anyone else could frontrun and use the users tokens as their input?

Reply:
Imo we might need a `burnFrom` to make this work, with approve/permit 🤪

Reply:
I think user facing routers should never really need claims?

Reply:
Yep agree, adressed offline that this is intended behavior